lecture-2023-10-05-cli-args/src/types/action/mod.rs

```diff
@@ -0,0 +1,102 @@
+use std::error::Error;
+use std::fmt;
+use std::fmt::Formatter;
+use std::str::FromStr;
+
+use strum::IntoEnumIterator;
+use strum_macros::EnumIter;
+
+
+/// `Action` represent all the available actions over the given input string.
+#[derive(PartialEq, Eq, Hash, EnumIter)]
+pub enum Action {
+    LowerCase,
+    UpperCase,
+    NoSpaces,
+    Slugify,
+    Palindromize,
+    MixedCase{first_lowercase: bool},
+    SwapPairs,
+    Caesar,
+    Csv,
+}
+
+
+impl fmt::Display for Action {
+    /// `fmt` enable conversion from `Action` to `&str` via `to_string` method.
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let key = match self {
+            Action::LowerCase => "lowercase",
+            Action::UpperCase => "uppercase",
+            Action::NoSpaces => "no-spaces",
+            Action::Slugify => "slugify",
+            Action::Palindromize => "palindromize",
+            Action::MixedCase {first_lowercase: true} => "lo-mixedcase",
+            Action::MixedCase {first_lowercase: false} => "hi-mixedcase",
+            Action::SwapPairs => "swap-pairs",
+            Action::Caesar => "caesar",
+            Action::Csv => "csv"
+        };
+
+        write!(f, "{}", key)
+    }
+}
+
+
+impl FromStr for Action {
+    type Err = String;
+
+    /// `from_str` implement conversion from `&str` to `Action` that might fail on invalid text.
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        let key = s.trim().to_lowercase();
+        let key = key.as_str();
+
+        // chain hack - due to not having there parameterized enum variant with non-default value
+        let action_list = Action::iter().chain([Action::MixedCase {first_lowercase: true}]);
```

Comment:
Haha, nice. This could be solved by having

Reply:
Could you, please, provide an example what you meant? I did not catch that _having_...